benchmark/benchmark.py

```diff
@@ -0,0 +1,534 @@
+            float(loss),
+            float(accuracy),
+            int(gpu_count),
+            gpu_model,
+        )
+
+    def rt_mms(self):
+        return self.run_time.total_seconds()
+
+    def write_row(self, writer: csv.DictWriter):
+        values = {x: None for x in RunData.CSV_FIELDS}
+        values[RunData.CSV_FIELDS[0]] = str(self.start_time)
+        values[RunData.CSV_FIELDS[1]] = str(self.end_time)
+        values[RunData.CSV_FIELDS[2]] = str(self.rt_mms())
+        values[RunData.CSV_FIELDS[3]] = self.model
+        values[RunData.CSV_FIELDS[4]] = self.parallel
+        values[RunData.CSV_FIELDS[5]] = self.epochs
+        values[RunData.CSV_FIELDS[6]] = self.batch_size
+        values[RunData.CSV_FIELDS[7]] = self.loss
+        values[RunData.CSV_FIELDS[8]] = self.accuracy
+        values[RunData.CSV_FIELDS[9]] = self.gpu_count
+        values[RunData.CSV_FIELDS[10]] = self.gpu_model
+        writer.writerow(values)
+
+    @staticmethod
+    def write_header(fp) -> csv.DictWriter:
+        writer = csv.DictWriter(fp, dialect="unix", fieldnames=RunData.CSV_FIELDS)
+        writer.writeheader()
+        return writer
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}({self.start_time}, {self.end_time}, {self.model}"
+            f", {self.parallel}, {self.epochs}, {self.batch_size}, {self.gpu_model}, {self.gpu_count}, {self.accuracy})"
+        )
+
+    def same(self, o: "RunData") -> bool:
+        return all([
+            self.model == o.model,
+            self.parallel == o.parallel,
+            self.batch_size == o.batch_size,
+            self.gpu_count == o.gpu_count,
+            self.gpu_model == o.gpu_model,
+        ])
+
+    def gt(self, o: "RunData") -> bool:
+        return self.end_time > o.end_time
+
+
+@dataclasses.dataclass
+class RunResult:
+    start_time: datetime.datetime
+    end_time: datetime.datetime
+    loss: float
+    accuracy: float
+
+
+BATCH_SIZE_VALUES = (1, 2, 4, 8, 16, 32, 64)
+
+# https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html
+# https://matplotlib.org/stable/gallery/lines_bars_and_markers/linestyles.html
+PLOT_LINES = [
+    "solid",
+    "dashed",
+    "dotted",
+    "dashdot",
+]
+PLOT_MARKERS = [
+    "x",
+    "+",
+    ".",
+    "*",
+    "H",
+    "s",
+]
+PLOT_IMG_WIDTH = 10
+PLOT_IMG_HEIGHT = 5
+PLOT_IMG_MARGIN_LEFT = 0.1
+
+
+def gen_plot_lines():
+    for e in PLOT_LINES:
+        yield e
+
+
+def gen_plot_markers():
+    for e in PLOT_MARKERS:
+        yield e
+
+
+def device_name(model: str, count: int) -> str:
+    return f"{count} {model}"
+
+
+def device_filename(model: str, count: int) -> str:
+    return f"{model} {count}".lower().replace(" ", "-")
+
+
+def nofn(args):
+    return 0, 0
+
+
+def wrap(fn, args, q):
+    start_time = datetime.datetime.now()
+    loss, accuracy = fn(args)
+    end_time = datetime.datetime.now()
+    q.put(RunResult(start_time, end_time, loss, accuracy))
```

Comment:
Please add proper error handling; if the training function raises an exception, it will terminate the subprocess, but the main process might hang indefinitely.

Reply:
I think the philosophy of Giotto-deep is to be as stable as possible; this requires proper error handling without the user figuring out what went wrong somewhere deep in the code.